### PR TITLE
New configuration to limit the arena extension

### PR DESCRIPTION
**Pull request by guoyuhong**
### Description

Add a new configuration `max_power_of_two_extend_bytes ` to limit the arena extension size.

### Motivation and Context

In our real scenario, we observe that if the model is big enough the BfcArena will extend uncontrollable. 
As showed by the following figures, if a model uses more than 16GB memory, the BfcArena will totally apply for 32GB memory according to the `kNextPowerOfTwo` strategy. With the new strategy, the extension is limited. The default maximum extension size is 1GB.

#### Without the new configuration
After loading the model, ORT uses 32G GPU memory.
image

#### With the new configuration
After loading the model, ORT uses 23G GPU memory.
image

### CODE REVIEW

**Comment by guoyuhong** [comment]
@tianleiwu Could you take a look?

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 1]
```diff
@@ -2579,7 +2581,7 @@ struct OrtApi {
    * This will create the configuration of an arena that can eventually be used to define an arena based allocator's behavior
    *
    * \param[in] max_mem Use 0 to allow ORT to choose the default
-   * \param[in] arena_extend_strategy Use -1 to allow ORT to choose the default, 0 = kNextPowerOfTwo, 1 = kSameAsRequested
+   * \param[in] arena_extend_strategy Use -1 to allow ORT to choose the default, 0 = kNextPowerOfTwo, 1 = kSameAsRequested, 2 = kLimitedNextPowerOfTwo
```
Comment (souptc) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:2584]: do we really need a new strategy? it looks more like adding a configuration to power of 2 strategy.

File: onnxruntime/core/session/environment.cc [hunk 2]
```diff
@@ -135,14 +135,15 @@ Status Environment::CreateAndRegisterAllocator(const OrtMemoryInfo& mem_info, co
...
     if (arena_cfg) {
       max_mem = arena_cfg->max_mem;
 
       arena_extend_strategy = arena_cfg->arena_extend_strategy;
       // validate the value here
-      if (!(arena_extend_strategy == -1 || arena_extend_strategy == 0 || arena_extend_strategy == 1)) {
+      if (!(arena_extend_strategy == -1 || arena_extend_strategy == 0 || arena_extend_strategy == 1 || arena_extend_strategy == 2)) {
         return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT,
                                "Received invalid value for arena extend strategy."
                                " Valid values can be either 0, 1 or -1.");
```
Comment (tianleiwu) [inline onnxruntime/core/session/environment.cc:149]: update message

File: include/onnxruntime/core/session/onnxruntime_c_api.h (hunk 1 above)
Comment (tianleiwu) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:2584]: It's fine to add a config only.
Note that python API does not expose the configuration max_limited_power_of_two_bytes.

Please update the documentation about the design change. Note that max_limited_power_of_two_bytes is not actual limit. When requested bytes is 2GB, allocator will expand 2GB. It is just a limit for expand when requested bytes < 1GB.

Comment (pranavsharma) [reply include/onnxruntime/core/session/onnxruntime_c_api.h:2584]: +1. No new strategy needed. Just add a config.

Comment (guoyuhong) [reply include/onnxruntime/core/session/onnxruntime_c_api.h]: Thanks for all your comments. I have removed the new strategy but only used the new configuration. The python API has expose the new configuration. 
@tianleiwu I'm not sure whether I have updated all the documentations. Please take a look~

File: onnxruntime/core/session/environment.cc [hunk 3]
```diff
@@ -135,14 +135,15 @@ Status Environment::CreateAndRegisterAllocator(const OrtMemoryInfo& mem_info, co
...
+    int64_t max_power_of_two_extend_bytes = -1L;
 
     // override with values from the user supplied arena_cfg object
     if (arena_cfg) {
       max_mem = arena_cfg->max_mem;
 
       arena_extend_strategy = arena_cfg->arena_extend_strategy;
       // validate the value here
-      if (!(arena_extend_strategy == -1 || arena_extend_strategy == 0 || arena_extend_strategy == 1)) {
+      if (!(arena_extend_strategy == -1 || arena_extend_strategy == 0 || arena_extend_strategy == 1 || arena_extend_strategy == 2)) {
```
Comment (tianleiwu) [inline onnxruntime/core/session/environment.cc:146]: undo this line

Comment (guoyuhong) [reply onnxruntime/core/session/environment.cc]: Good catch~Done.

**Review by tianleiwu — DISMISSED** [review]

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**